src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java

```diff
@@ -171,35 +174,47 @@ public LectureResource(LectureRepository lectureRepository, LectureService lectu
         if (!authCheckService.isAtLeastStudentInCourse(course, user)) {
             return forbidden();
         }
+        lecture = filterLectureContentForUser(lecture, user);
+
+        return ResponseEntity.ok(lecture);
+    }
+
+    private Lecture filterLectureContentForUser(Lecture lecture, User user) {
         lecture = lectureService.filterActiveAttachments(lecture, user);
 
-        // to make sure that the information provided for lecture units is equal to the one in the course dashboard
-        // ToDo Improve performance by constructing a more precise call
-        course = this.courseResource.getCourseForDashboard(course.getId());
-        Set<Exercise> exercisesUserIsAllowedToSee = course.getExercises();
+        // The Objects::nonNull is needed here because the relationship lecture -> lecture units is ordered and
+        // hibernate sometimes adds nulls to in the list of lecture units to keep the order
+        Set<Exercise> relatedExercises = lecture.getLectureUnits().stream().filter(Objects::nonNull).filter(lectureUnit -> lectureUnit instanceof ExerciseUnit)
+                .map(lectureUnit -> ((ExerciseUnit) lectureUnit)).map(ExerciseUnit::getExercise).collect(Collectors.toSet());
```

Comment:
Can't this simplify to a single map?
`.map(lectureUnit -> ((ExerciseUnit) lectureUnit).getExercise())`

Reply:
instanceof includes null checks, so the nonNull might not be necessary because `null instanceof ExerciseUnit` would evaluate to false, but it does not hurt ;-)

Reply:
I agree that I could remove the extra null check but I think in this case we should keep it in so if this method is every changed in the future, we don't run into any problems. The performance cost should be very low

Reply:
This was actually not about the null check but rather about converting to an `ExerciseUnit` and getting the exercise from that in one instead of two steps.

Reply:
Done in 56436adaffe06cfe551221f2d7314a6a2194278f